src/modules/auth/services/users.service.spec.ts

```diff
@@ -0,0 +1,165 @@
+import { Test } from '@nestjs/testing';
+import { UsersService } from './users.service';
+import { User } from '../entities/user.entity';
+import { getRepositoryToken } from '@nestjs/typeorm';
+import * as bcrypt from 'bcrypt';
+
+import type { UpdateUserDto } from '../dtos/update.user.dto';
+import type { CreateUserDto } from '../dtos/createUser.dto';
+import { BadRequestException, ForbiddenException, NotFoundException } from '@nestjs/common';
+type findSchema = {
+  where: {
+    email?: string;
+    id?: number;
+  };
+};
+
+describe('UsersService', () => {
+  let service: UsersService;
+
+  const mockUser = {
+    id: 1,
+    email: 'example@mail.com',
+    first_name: 'John',
+    last_name: 'Doe',
+    password: bcrypt.hashSync('password', 10),
+    image: Buffer.from('', 'base64'),
+    hashedRt: '',
+  };
+
+  const mockRepository = {
+    findOne: jest.fn(async ({ where }: findSchema) => {
+      if (where.id === mockUser.id || where.email === mockUser.email) {
+        return await mockUser;
+      }
+
+      return null;
+    }),
+    create: jest.fn(async (dto: CreateUserDto) => {
+      return await {
+        ...dto,
+        id: 2,
+        hashedRt: null,
+      };
+    }),
+    save: jest.fn(async (user: User) => {
+      return await user;
+    }),
+    update: jest.fn(async (id: number, updateUserDto: UpdateUserDto) => {
+      if (updateUserDto && id !== mockUser.id) {
+        return await { affected: 0 };
+      }
+      return await { affected: 1 };
+    }),
+    delete: jest.fn(async (id: number) => {
+      if (id !== mockUser.id) {
+        return await { affected: 0 };
+      }
+      return await { affected: 1 };
+    }),
+  };
+
+  beforeEach(async () => {
+    const module = await Test.createTestingModule({
+      providers: [
+        UsersService,
+        { provide: getRepositoryToken(User), useValue: mockRepository },
+      ],
+    }).compile();
+
+    service = module.get<UsersService>(UsersService);
+  });
```

Comment:
```suggestion
        { provide: UsersService, useClass: UsersServiceImpl },
        { provide: UserRepository, useValue: mockUserRepository },
      ],
    }).compile();

    userService = module.get<UsersService>(UsersService);
  });
```

UsersService - абстрактний клас, який описує клас імплементацію
UsersServiceImpl - клас імплементація
UsersServiceMock - клас який імплементує UsersService (для інших тестів де юзатиметься цей сервіс)

Reply:
Також окремо файл репозиторію

import { Injectable } from '@nestjs/common';
import { DataSource, Repository } from 'typeorm';
import User from '../entities/User.entity';

@Injectable()
export class UserRepository extends Repository<User> {

  constructor(_dataSource: DataSource) {
    super(User, _dataSource.createEntityManager());
  }

}